imputation/ts_impute.go

```diff
@@ -50,42 +59,102 @@ func dateGapFinder(keys []string) (string, int, int, int) {
 
 	year, month, day := 0, 0, 0
 
-	duration := 1<<31 - 1 // A large number.
+	equalMonths := true
+	equalDays := true
+
+	if parseFormat != yearfmt {
+		for i := 0; i < len(keys)-1; i++ {
+			md1 := strings.Split(keys[i], "-")
+			md2 := strings.Split(keys[i+1], "-")
+			// Check for unequal months.
+			if md1[1] != md2[1] {
+				equalMonths = false
+			}
+			// Check for unequal days (if exists).
+			if len(md1) == 3 {
+				if md1[2] != md2[2] {
+					equalDays = false
+				}
+			}
+		}
+	}
 
-	for i := 0; i < len(keys)-1; i++ {
-		start, _ := time.Parse(parseFormat, keys[i])
-		end, _ := time.Parse(parseFormat, keys[i+1])
+	if parseFormat == yearfmt || (equalMonths && equalDays) {
+		for i := 0; i < len(keys)-1; i++ {
+			if year == 1 {
+				// One year is the minimum possible gap in this case.
+				break
+			}
+			start, _ := time.Parse(parseFormat, keys[i])
+			end, _ := time.Parse(parseFormat, keys[i+1])
 
-		// Calculate total number of days between each two points and compare to minimum gap so far.
-		if delta := int(end.Sub(start).Hours() / 24); duration > delta { //TODO: instead of min use GCD
-			duration = delta
-			y1, M1, d1 := start.Date()
-			y2, M2, d2 := end.Date()
+			y1, _, _ := start.Date()
+			y2, _, _ := end.Date()
+			yGap := int(y2 - y1)
 
-			year = int(y2 - y1)
-			month = int(M2 - M1)
-			day = int(d2 - d1)
+			if i == 0 {
+				// Initialize year with the value of the first gap.
+				year = yGap
+			} else {
+				year = gcd(yGap, year)
+			}
 		}
-	}
+	} else if parseFormat == monthfmt || equalDays {
+		// One month is the minimum possible gap in this case.
+		for i := 0; i < len(keys)-1; i++ {
+			if month == 1 {
+				break
+			}
+			start, _ := time.Parse(parseFormat, keys[i])
+			end, _ := time.Parse(parseFormat, keys[i+1])
 
-	return parseFormat, year, month, day
-}
+			y1, M1, _ := start.Date()
+			y2, M2, _ := end.Date()
+			yGap := int(y2 - y1)
+			mGap := int(M2 - M1)
+			if mGap < 1 {
+				mGap += 12
+				yGap -= 1
+			}
 
-func getSortedKeys(ts TimeSeries) []string {
-	keys := make([]string, 0)
-	for k, _ := range ts {
-		keys = append(keys, k)
+			if i == 0 {
+				// Initialize year and month with the value of the first gap.
+				year = yGap
+				month = mGap
+			} else {
+				log.Println(mGap)
+				month = gcd(mGap, month)
```

Comment:
I do not think this computes the correct value for a series like "2012-01", "2013-02", "2014-03", "2015-04" for which I think the value should be 13 months, not 1 month.  If you switch to year*12+month, the problem disappears.